cirq/protocols/unitary.py

```diff
@@ -0,0 +1,88 @@
+# Copyright 2018 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Any, Optional, TypeVar, Union
+
+import numpy as np
+from typing_extensions import Protocol
+
+from cirq import abc, extension
+from cirq.ops.gate_features import KnownMatrix
+
+# This is a special indicator value used by the unitary method to determine
+# whether or not the caller provided a 'default' argument. It must be of type
+# np.ndarray to ensure the method has the correct type signature in that case.
+# It is checked for using `is`, so it won't have a false positive if the user
+# provides a different np.array([]) value.
+RaiseTypeErrorIfNotProvided = np.array([])
+
+TDefault = TypeVar('TDefault')
+
+
+class SupportsUnitary(Protocol):
```

Comment:
Do we want to support runtime isinstance checks of protocols? https://mypy.readthedocs.io/en/latest/protocols.html#using-isinstance-with-protocols

Reply:
Mostly I wanted mypy to understand the type. But yes I think it makes sense for the runtime check to work.